core/src/main/scala/org/typelevel/otel4s/trace/Span.scala

```diff
@@ -0,0 +1,225 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package trace
+
+import scala.concurrent.duration.FiniteDuration
+
+/** The API to trace an operation.
+  *
+  * There are three types of span: [[Span.Manual]], [[Span.Auto]], [[Span.Res]].
+  *
+  * ==[[Span.Manual]]==
+  * The manual span requires to be ended '''explicitly'''. Manual span can be
+  * used when it's necessary to end a span outside of the resource scope (i.e.
+  * async callback). Make sure the span is ended properly.
+  *
+  * Leaked span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val leaked: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createManual.use { span =>
+  *     span.setStatus(Status.Ok, "all good")
+  *   }
+  * }}}
+  *
+  * Properly ended span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createManual.use { span =>
+  *     span.setStatus(Status.Ok, "all good") >> span.end
+  *   }
+  * }}}
+  *
+  * ==[[Span.Auto]]==
+  * Unlike [[Span.Manual]] the auto span has a fully managed lifecycle. That
+  * means the span is started upon resource allocation and ended upon
+  * finalization.
+  *
+  * Automatically ended span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createAuto.use { span =>
+  *     span.setStatus(Status.Ok, "all good")
+  *   }
+  * }}}
+  *
+  * ==[[Span.Res]]==
+  * The behaviour and lifecycle management are identical to [[Span.Auto]]. The
+  * allocation and release stages of a supplied resource are traced by separate
+  * spans. Carries a value of a wrapped resource.
+  *
+  * The structure of the inner spans:
+  * {{{
+  * > span-name
+  *   > acquire
+  *   > use
+  *   > release
+  * }}}
+  *
+  * Resource span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val resource: Resource[F, String] = Resource.eval(Sync[F].delay("string"))
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createRes(resource).use { case span @ Span.Res(value) =>
+  *     span.setStatus(Status.Ok, s"all good. resource value: $${value}")
+  *   }
+  * }}}
+  */
+trait Span[F[_]] {
+
+  /** Returns the [[SpanContext]] associated with this span.
+    */
+  def context: SpanContext
+
+  /** Sets attributes to the span. If the span previously contained a mapping
+    * for any of the keys, the old values are replaced by the specified values.
+    *
+    * @param attributes
+    *   the set of attributes to add to the span
+    */
```

Comment:
This is getting pretty specific to the spec, but the Otel spec requires the API to provide a single attribute setter, but then optionally a setter for multiple attributes at once.

Reply:
Good question. I'm fine with varargs. 
`span.setAttributes(Attribute(...), Attribute(...))` is more handy than `span.setAttributes(Seq(Attribute(...), Attribute(...)))`

We still can add `def setAttribute[A](attribute: Attribute[A]): F[Unit]`, it will not hurt for sure.

Reply:
Yeah I think addiing `setAttribute` just to fulfill the requirement and make it easy to communicate about api shapes to other folks using Otel would be great.

Reply:
I added the method

Reply:
Thanks!  Looks great.